<h1>Содержание<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Подключение-к-базе-данных" data-toc-modified-id="Подключение-к-базе-данных-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Подключение к базе данных</a></span></li><li><span><a href="#Исследуйте-таблицы-—-выведите-первые-строки" data-toc-modified-id="Исследуйте-таблицы-—-выведите-первые-строки-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Исследуйте таблицы — выведите первые строки</a></span></li><li><span><a href="#Посчитайте,-сколько-книг-вышло-после-1-января-2000-года" data-toc-modified-id="Посчитайте,-сколько-книг-вышло-после-1-января-2000-года-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Посчитайте, сколько книг вышло после 1 января 2000 года</a></span></li><li><span><a href="#Для-каждой-книги-посчитайте-количество-обзоров-и-среднюю-оценку" data-toc-modified-id="Для-каждой-книги-посчитайте-количество-обзоров-и-среднюю-оценку-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Для каждой книги посчитайте количество обзоров и среднюю оценку</a></span></li><li><span><a href="#Определите-издательство,-которое-выпустило-наибольшее-число-книг-толще-50-страниц-—-так-вы-исключите-из-анализа-брошюры" data-toc-modified-id="Определите-издательство,-которое-выпустило-наибольшее-число-книг-толще-50-страниц-—-так-вы-исключите-из-анализа-брошюры-5"><span class="toc-item-num">5&nbsp;&nbsp;</span>Определите издательство, которое выпустило наибольшее число книг толще 50 страниц — так вы исключите из анализа брошюры</a></span></li><li><span><a href="#Определите-автора-с-самой-высокой-средней-оценкой-книг-—-учитывайте-только-книги-с-50-и-более-оценками" data-toc-modified-id="Определите-автора-с-самой-высокой-средней-оценкой-книг-—-учитывайте-только-книги-с-50-и-более-оценками-6"><span class="toc-item-num">6&nbsp;&nbsp;</span>Определите автора с самой высокой средней оценкой книг — учитывайте только книги с 50 и более оценками</a></span></li><li><span><a href="#Посчитайте-среднее-количество-обзоров-от-пользователей,-которые-поставили-больше-50-оценок" data-toc-modified-id="Посчитайте-среднее-количество-обзоров-от-пользователей,-которые-поставили-больше-50-оценок-7"><span class="toc-item-num">7&nbsp;&nbsp;</span>Посчитайте среднее количество обзоров от пользователей, которые поставили больше 50 оценок</a></span></li></ul></div>

# Проект по SQL

Во время короновируса в свободное время жители городов стали больше читать различные книгни. Данный факт не остался без внимания стартаперов и они ринулись создавать приложения для читения книг. Наша компания купила крупный сервис для чтения книг по подписке. Необходимо проанализировать базу данных. В ней содержиться информация о книгах, издательствах, авторах, а также пользовательские обзоры книг. Эти данные помогут улучшить новый продукт.

# Описание данных

**Таблица books**

Содержит данные о книгах:
- book_id — идентификатор книги;
- author_id — идентификатор автора;
- title — название книги;
- num_pages — количество страниц;
- publication_date — дата публикации книги;
- publisher_id — идентификатор издателя.

**Таблица authors**

Содержит данные об авторах:
- author_id — идентификатор автора;
- author — имя автора.

**Таблица publishers**

Содержит данные об издательствах:
- publisher_id — идентификатор издательства;
- publisher — название издательства;

**Таблица ratings**

Содержит данные о пользовательских оценках книг:
- rating_id — идентификатор оценки;
- book_id — идентификатор книги;
- username — имя пользователя, оставившего оценку;
- rating — оценка книги.

**Таблица reviews**

Содержит данные о пользовательских обзорах:
- review_id — идентификатор обзора;
- book_id — идентификатор книги;
- username — имя автора обзора;
- text — текст обзора.

## Подключение к базе данных

In [1]:
# импортируем библиотеки
import pandas as pd
from sqlalchemy import create_engine

In [2]:
# устанавливаем параметры
db_config = {'user': 'praktikum_student', # имя пользователя
'pwd': 'Sdf4$2;d-d30pp', # пароль
'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
'port': 6432, # порт подключения
'db': 'data-analyst-final-project-db'} # название базы данных
connection_string = 'postgresql://{}:{}@{}:{}/{}'.format(db_config['user'],
 db_config['pwd'],
 db_config['host'],
 db_config['port'],
 db_config['db'])

In [3]:
# сохраняем коннектор
engine = create_engine(connection_string, connect_args={'sslmode':'require'}) 

In [4]:
def select(query):
    return pd.read_sql(query,con=engine)

## Исследуйте таблицы — выведите первые строки

Выведем на экран первые 5 строк таблицы с данными о книгах

In [5]:
sql_books = ''' SELECT * FROM books '''
books = select(sql_books)
books.head()

book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268

Выведем на экран первые 5 строк таблицы с данными о авторах 

In [6]:
sql_authors = ''' SELECT * FROM authors '''
authors = select(sql_authors)
authors.head()

author_id                          author
0          1                      A.S. Byatt
1          2  Aesop/Laura Harris/Laura Gibbs
2          3                 Agatha Christie
3          4                   Alan Brennert
4          5        Alan Moore/David   Lloyd

Выведем на экран первые 5 строк таблицы с данными об издательствах

In [7]:
sql_publishers= ''' SELECT * FROM publishers '''
publishers = select(sql_publishers) 
publishers.head()

publisher_id                          publisher
0             1                                Ace
1             2                           Ace Book
2             3                          Ace Books
3             4                      Ace Hardcover
4             5  Addison Wesley Publishing Company

Выведем на экран первые 5 строк таблицы с данными о пользовательских оценках книг 

In [8]:
sql_ratings = ''' SELECT * FROM ratings '''
ratings = select(sql_ratings) 
ratings.head()

rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5
3          4        2       lorichen       3
4          5        2    mariokeller       2

Выведем на экран первые 5 строк таблицы с данными о пользовательских обзорах на книги

In [9]:
sql_reviews = ''' SELECT * FROM reviews '''
reviews = select(sql_reviews) 
reviews.head()

review_id  book_id       username  \
0          1        1   brandtandrea   
1          2        1     ryanfranco   
2          3        2       lorichen   
3          4        3  johnsonamanda   
4          5        3    scotttamara   

                                                text  
0  Mention society tell send professor analysis. ...  
1  Foot glass pretty audience hit themselves. Amo...  
2  Listen treat keep worry. Miss husband tax but ...  
3  Finally month interesting blue could nature cu...  
4  Nation purpose heavy give wait song will. List...

## Посчитайте, сколько книг вышло после 1 января 2000 года

In [10]:
query = '''
SELECT COUNT(DISTINCT book_id) AS count_books
FROM books
WHERE publication_date >= '2000-01-01'
'''
select(query)

count_books
0          821

После 1 января 2000 года выпущена 821 книга

## Для каждой книги посчитайте количество обзоров и среднюю оценку

In [11]:
query = '''
SELECT
 books.book_id AS id,
 books.title AS title_book,
 AVG(ratings.rating) AS AVG_rating,
 COUNT(DISTINCT reviews.review_id) AS number_of_reviews
FROM books
 INNER JOIN reviews ON reviews.book_id = books.book_id
 INNER JOIN ratings ON ratings.book_id = books.book_id
GROUP BY
 id,
 title_book
ORDER BY
 number_of_reviews DESC
LIMIT 10
'''
select(query)

id                                         title_book  avg_rating  \
0  948                            Twilight (Twilight  #1)    3.662500   
1  696             The Da Vinci Code (Robert Langdon  #2)    3.830508   
2  207                                    Eat  Pray  Love    3.395833   
3  627                                      The Alchemist    3.789474   
4  673                             The Catcher in the Rye    3.825581   
5  695  The Curious Incident of the Dog in the Night-Time    4.081081   
6  302  Harry Potter and the Prisoner of Azkaban (Harr...    4.414634   
7  299  Harry Potter and the Chamber of Secrets (Harry...    4.287500   
8  497                          Outlander (Outlander  #1)    4.125000   
9  656                                     The Book Thief    4.264151   

   number_of_reviews  
0                  7  
1                  6  
2                  6  
3                  6  
4                  6  
5                  6  
6                  6  
7                  6  
8                  6  
9                  6

Вывели топ 10 книг по количеству обзоров. Лидером оказалась книга "Twilight (Twilight #1)" у нее 7 обзоров, хотя и не самый большой ретинг (3,7)

## Определите издательство, которое выпустило наибольшее число книг толще 50 страниц — так вы исключите из анализа брошюры

In [12]:
query = '''
SELECT
 publishers.publisher_id AS id,
 publishers.publisher AS publisher,
 COUNT(DISTINCT books.book_id) AS number_of_books
FROM publishers
 LEFT JOIN books ON books.publisher_id = publishers.publisher_id
WHERE 
 books.num_pages > 50
GROUP BY
 id,
 publisher
ORDER BY
 number_of_books DESC
LIMIT 10
'''
select(query)

id                 publisher  number_of_books
0  212             Penguin Books               42
1  309                   Vintage               31
2  116  Grand Central Publishing               25
3  217          Penguin Classics               24
4   33          Ballantine Books               19
5   35                    Bantam               19
6   45                   Berkley               17
7   46             Berkley Books               14
8  284        St. Martin's Press               14
9   83                     Delta               13

Оказалось, что издательство выпустившее наибольшее количество книг - это Penguin Books

## Определите автора с самой высокой средней оценкой книг — учитывайте только книги с 50 и более оценками

In [13]:
query = '''
WITH data as (
SELECT
   a.author,
   r.book_id,
   COUNT(rating) as count_rating,
   AVG(rating) as avg_rating
FROM
   ratings r
LEFT JOIN books b ON r.book_id = b.book_id
LEFT JOIN authors a ON b.author_id = a.author_id
GROUP BY
    a.author,
    r.book_id
HAVING
   COUNT(rating) > 50
   )

SELECT
    author,
    AVG(avg_rating) as avg_rating,
    SUM(count_rating)
FROM
    data
GROUP BY
    author
ORDER BY
    avg_rating DESC
'''
select(query)

author  avg_rating    sum
0                          J.K. Rowling/Mary GrandPré    4.283844  310.0
1                   Markus Zusak/Cao Xuân Việt Khương    4.264151   53.0
2                                      J.R.R. Tolkien    4.258446  162.0
3                                   Louisa May Alcott    4.192308   52.0
4                                        Rick Riordan    4.080645   62.0
5                                     William Golding    3.901408   71.0
6                                       J.D. Salinger    3.825581   86.0
7            Paulo Coelho/Alan R. Clarke/Özdemir İnce    3.789474   57.0
8   William Shakespeare/Paul Werstine/Barbara A. M...    3.787879   66.0
9                                           Dan Brown    3.754540  143.0
10                                         Lois Lowry    3.750000   56.0
11            George Orwell/Boris Grabnar/Peter Škerl    3.729730   74.0
12                                    Stephenie Meyer    3.662500  160.0
13                                     John Steinbeck    3.622951   61.0

Автор с самой высокой среденей оценкой книг - J.K. Rowling/Mary GrandPré, за ней идёт Markus Zusak/Cao Xuân Việt Khương

## Посчитайте среднее количество обзоров от пользователей, которые поставили больше 50 оценок

In [14]:
query = '''
SELECT
 AVG (number_of_reviews) AS AVG_count
 FROM
 (
   SELECT
    COUNT (review_id) AS number_of_reviews
   FROM
    reviews
   INNER JOIN
   (SELECT
     username,
     COUNT (rating_id) AS number_of_ratings
    FROM
     ratings
    GROUP BY
     username
    HAVING COUNT (rating_id) > 50 ) AS ratings_users ON ratings_users.username = reviews.username
    GROUP BY
    reviews.username
 ) AS number_of_reviews
'''
select(query)

avg_count
0  24.333333

Среднее количество обзоров от пользователей, которые поставили больше 50 оценок равно 24

# Итоги

- После 1 января 2000 года выпущена 821 книга
- Лидером среди книг по количеству обзоров оказалась книга "Twilight (Twilight #1)" у нее 7 обзоров
- Издательство выпустившее наибольшее количество книг - это Penguin Books
- Автор с самой высокой среденей оценкой книг - J.K. Rowling/Mary GrandPré
- Среднее количество обзоров от пользователей, которые поставили больше 50 оценок равно 24